From a3cab40bc2d1b0e83e22f61e44f6c18f0fa285e6 Mon Sep 17 00:00:00 2001
From: dineshsv <dineshsv@ami.com>
Date: Wed, 26 Jun 2024 14:52:14 +0530
Subject: [PATCH] Added SEL Enable/Disable via Set BMC Global Enables IPMI
 Command

Tested :

 - Verified by Disabling SEL and triggerred sensor event,
   Not able to log the Event in DBUS.

 - Working as expexted.

```shell

root@AMIOT-d6490714d8c2:~# busctl get-property   xyz.openbmc_project.Settings /xyz/openbmc_project/control/globalenables xyz.openbmc_project.Control.BMC.Globalenables Sel

b false

xyz.openbmc_project.Control.BMC.Globalenables interface -         -            -
.EventmsgBuf                                  property  b         false        emits-change writable
.EventmsgFullintr                             property  b         false        emits-change writable
.OEM0                                         property  b         false        emits-change writable
.OEM1                                         property  b         false        emits-change writable
.OEM2                                         property  b         false        emits-change writable
.RecvMsgQueueInterrupt                        property  b         true         emits-change writable
.Sel                                          property  b         false        emits-change writable

root@AMIOT-d6490714d8c2:~# ipmitool sel list
   1 | 06/03/24 | 10:01:15 UTC | Temperature #0x30 | Upper Critical going high | Asserted

```

Triggered sensor event:

```shell

root@AMIOT-d6490714d8c2:~# ipmitool event 2
Sending SAMPLE event: Voltage Threshold - Lower Critical - Going Low
   0 |  Pre-Init  |0000000000| Voltage #0x60 | Lower Critical going low  | Asserted

root@AMIOT-d6490714d8c2:~# ipmitool sel list
   1 | 06/03/24 | 10:01:15 UTC | Temperature #0x30 | Upper Critical going high | Asserted

root@AMIOT-d6490714d8c2:~# ipmitool sel list
   1 | 06/03/24 | 10:01:15 UTC | Temperature #0x30 | Upper Critical going high | Asserted

root@AMIOT-d6490714d8c2:~# ipmitool sel list
   1 | 06/03/24 | 10:01:15 UTC | Temperature #0x30 | Upper Critical going high | Asserted

root@AMIOT-d6490714d8c2:~# busctl tree xyz.openbmc_project.Logging
`- /xyz
  `- /xyz/openbmc_project
    `- /xyz/openbmc_project/logging
      |- /xyz/openbmc_project/logging/entry
      | `- /xyz/openbmc_project/logging/entry/1
      `- /xyz/openbmc_project/logging/internal
        `- /xyz/openbmc_project/logging/internal/manager

```

Signed-off-by: dineshsv <dineshsv@ami.com>
---
 log_manager.cpp | 33 +++++++++++++++++++++++++++++++++
 1 file changed, 33 insertions(+)

diff --git a/log_manager.cpp b/log_manager.cpp
index a5c7955..f711e16 100644
--- a/log_manager.cpp
+++ b/log_manager.cpp
@@ -13,6 +13,7 @@
 #include <unistd.h>
 
 #include <phosphor-logging/lg2.hpp>
+#include <sdbusplus/bus.hpp>
 #include <sdbusplus/vtable.hpp>
 #include <xyz/openbmc_project/State/Host/server.hpp>
 
@@ -28,6 +29,7 @@
 #include <set>
 #include <string>
 #include <string_view>
+#include <variant>
 #include <vector>
 
 using namespace std::chrono;
@@ -45,6 +47,12 @@ constexpr auto policyLinear =
 constexpr auto policyDefault =
     "xyz.openbmc_project.Logging.Settings.Policy.Circular";
 
+/*D-bus details for BMC Global enable */
+static constexpr const char* settingService = "xyz.openbmc_project.Settings";
+static constexpr const char* globalEnblObjpath =
+    "/xyz/openbmc_project/control/globalenables";
+static constexpr const char* globalEnblInterface =
+    "xyz.openbmc_project.Control.BMC.Globalenables";
 using DBusInterface = std::string;
 using DBusService = std::string;
 using DBusPath = std::string;
@@ -257,6 +265,31 @@ void Manager::createEntry(std::string errMsg, Entry::Level errLvl,
                           std::vector<std::string> additionalData,
                           const FFDCEntries& ffdc)
 {
+    bool selEnabled;
+    sdbusplus::bus::bus bus = sdbusplus::bus::new_default();
+    auto method = bus.new_method_call(settingService, globalEnblObjpath,
+                                      "org.freedesktop.DBus.Properties", "Get");
+    // Append the interface and property name to the method call
+    method.append(globalEnblInterface, "Sel");
+
+    try
+    {
+        auto reply = bus.call(method);
+        // Extract the value from the response
+        std::variant<bool> value;
+        reply.read(value);
+        selEnabled = std::get<bool>(value);
+    }
+    catch (const sdbusplus::exception_t& e)
+    {
+        lg2::error("Failed to get D-Bus property:{ERROR}", "ERROR", e);
+        selEnabled = false;
+    }
+    if (!selEnabled)
+    {
+        lg2::info("SEL is disabled");
+        return;
+    }
     if (!Extensions::disableDefaultLogCaps())
     {
         std::string currentPolicy = getSelPolicy();
-- 
2.25.1

